tests/models/test_ip_adapters.py

```diff
@@ -0,0 +1,415 @@
+# coding=utf-8
+# Copyright 2023 HuggingFace Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import gc
+import unittest
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from transformers import (
+    CLIPImageProcessor,
+    CLIPTextConfig,
+    CLIPTextModel,
+    CLIPTokenizer,
+    CLIPVisionConfig,
+    CLIPVisionModelWithProjection,
+)
+
+from diffusers import (
+    AutoencoderKL,
+    DDIMScheduler,
+    StableDiffusionImg2ImgPipeline,
+    StableDiffusionInpaintPipeline,
+    StableDiffusionPipeline,
+    StableDiffusionXLImg2ImgPipeline,
+    StableDiffusionXLInpaintPipeline,
+    StableDiffusionXLPipeline,
+    UNet2DConditionModel,
+)
+from diffusers.models.attention_processor import (
+    AttnProcessor,
+    AttnProcessor2_0,
+    IPAdapterAttnProcessor,
+    IPAdapterAttnProcessor2_0,
+)
+from diffusers.models.embeddings import ImageProjection
+from diffusers.utils import load_image
+from diffusers.utils.testing_utils import (
+    enable_full_determinism,
+    nightly,
+    require_torch_gpu,
+    torch_device,
+)
+
+
+enable_full_determinism()
+
+
+class IPAdapterFastTests(unittest.TestCase):
```

Comment:
What I don't fully understand here is why we essentially test everything with pipelines here. IP adapters are just applied to the UNet model, so all we need to test here is that the adapter can be loaded into the UNet and then we can check that that the forward pass works correctly, that the unet can be saved and reloaded etc...

Reply:
it is a method on the pipeline no? so it is better to test on the pipeline? I see here for this particular test, we created a state_dict input (instead of a checkpint), so essentially we only tested on its capability to load into unet and run a forward pass.... so it could be done on unet instead

I think maybe we can make a "tiny-ip-adater" checkpoint and test with pipeline? if the part that applies to unet fails the test will fail anyway. 

slightly confused so let me know if I missed anything :)

Reply:
Let's talk about it later in the call here :-) 

Now that we have a `unet._load_ip_adapter...` method we should just add some tests to the unet